pydantic/_internal/_generate_schema.py

```diff
@@ -292,10 +296,18 @@ def _typed_dict_schema(self, typed_dict_cls: Any) -> core_schema.TypedDictSchema
         """
         Generate schema for a TypedDict.
         """
-        try:
-            required_keys: typing.FrozenSet[str] = typed_dict_cls.__required_keys__
-        except AttributeError:
-            raise TypeError('Please use `typing_extensions.TypedDict` instead of `typing.TypedDict`.')
+        # It is not possible to track required/optional keys in TypedDict without __required_keys__
+        # since TypedDict.__new__ erases the base classes (it replaces them with just `dict`)
+        # and thus we can track usage of total=True/False
+        # __required_keys__ was added in Python 3.9 (https://github.com/miss-islington/cpython/blob/1e9939657dd1f8eb9f596f77c1084d2d351172fc/Doc/library/typing.rst?plain=1#L1546-L1548)  # noqa: E501
+        # however it is buggy (https://github.com/python/typing_extensions/blob/ac52ac5f2cb0e00e7988bae1e2a1b8257ac88d6d/src/typing_extensions.py#L657-L666)  # noqa: E501
+        # Hence to avoid creating validators that do not do what users expect we only
+        # support typing.TypedDict on Python >= 3.11 or typing_extension.TypedDict on all versions
+        is_typing_typeddict = type(typed_dict_cls).__module__ == 'typing'
+        if not _SUPPORTS_TYPEDDICT and is_typing_typeddict:
```

Comment:
No big pushback but generally as soon as I have more than 1 boolean I like to give each condition a name so that the logic branch is more readable i.e. `if not we_support_typeddict_on_this_python_version and typed_dict_class_is_from_the_typing_module:` is more readable to me than `if sys.version_info < (3, 11) and type(typed_dict_cls).__module__ == 'typing':`. But that's a bit of personal preference, especially for 2 items. It think it is almost required if we're talking about 4+ items or combining and/or conditions.